### PR TITLE
Adding/updating logrotate configurations for agent and extension configurations.

### PR DESCRIPTION
**Pull request by vrdmr**
## Description
- Limiting the agent log files based on size.
- Adding a log rotation policy for extension logs.

Issue #1562, related to the work being done in #1714.

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1740)

### CODE REVIEW

File: config/waagent-extn.logrotate [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/var/log/azure/*/*.log {
+    # Old versions of log files are compressed with gzip by default.
+    compress
+
+    # Rotate log files > 10 MB, and keep last 20 archived files. With an average compression ratio of around 10:1, The
+    # archived files would be take an average of around 10 MB.
```
Comment (larohra) [inline config/waagent-extn.logrotate]: NIT: Typo here - ` archived files would be take an average of around 10 MB`

File: config/waagent.logrotate [hunk 2]
```diff
@@ -1,7 +1,22 @@
 /var/log/waagent.log {
+    # Old versions of log files are compressed with gzip by default.
     compress
-    monthly
-    rotate 6
+
+    # Rotate log files > 20 MB, and keep last 50 archived files. With a compression ratio of around 10:1, The archived
+    # files would be take an average of around 100 MB
```
Comment (larohra) [inline config/waagent.logrotate]: NIT: Typo here - `files would be take an average of around 100 MB`

File: config/waagent-extn.logrotate [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/var/log/azure/*/*.log {
+    # Old versions of log files are compressed with gzip by default.
+    compress
+
+    # Rotate log files > 10 MB, and keep last 20 archived files. With an average compression ratio of around 10:1, The
```
Comment (narrieta) [inline config/waagent-extn.logrotate]: > 20 archived files

10?

File: config/waagent.logrotate [hunk 4]
```diff
@@ -1,7 +1,22 @@
 /var/log/waagent.log {
+    # Old versions of log files are compressed with gzip by default.
     compress
-    monthly
-    rotate 6
+
+    # Rotate log files > 20 MB, and keep last 50 archived files. With a compression ratio of around 10:1, The archived
```
Comment (narrieta) [inline config/waagent.logrotate]: how did you get this ratio? thanks

File: config/waagent.logrotate [hunk 5]
```diff
@@ -1,7 +1,22 @@
...
+    # files would be take an average of around 100 MB
+    size 20M
```
Comment (narrieta) [inline config/waagent.logrotate]: how often are those files checked? I've seen logs > 1 GB... 50 may be too many

Comment (vrdmr) [reply config/waagent.logrotate]: `logrotate` runs once a day. The `logrotate's` cron file is kept in `/etc/cron.daily/`.

The 50 here would be the compressed files, and not uncompressed.

File: config/waagent.logrotate (hunk 4 above)
Comment (vrdmr) [reply config/waagent.logrotate]: This number is anecdotal, from trying different waagent.log files.

```
v@dev:/var/log$ ls -l waagent.log*
-rw-r--r-- 1 root root  9461 Dec 26 17:17 waagent.log
-rw-r--r-- 1 root root  5968 Dec  1 04:51 waagent.log.1.gz
-rw-r--r-- 1 root root 26143 Nov  1 05:39 waagent.log.2.gz
v@dev:/var/log$ gunzip -c waagent.log.1.gz | wc --bytes
57636
v@dev:/var/log$ gunzip -c waagent.log.2.gz | wc --bytes
285857
```

File: config/waagent-extn.logrotate (hunk 3 above)
Comment (vrdmr) [reply config/waagent-extn.logrotate]: The 20 comes from the Windows agent. They keep 20 files of the compressed rolled-over logs.

File: config/waagent.logrotate (hunk 5 above)
Comment (narrieta) [reply config/waagent.logrotate]: but still, how well do those huge files (> 1GB) compress? i think i have some samples from incidents i've looked at, let me see if i can find them

File: config/waagent-extn.logrotate (hunk 3 above)
Comment (narrieta) [reply config/waagent-extn.logrotate]: i meant line 8 says 10

Comment (narrieta) [reply config/waagent-extn.logrotate]: (plus the accounting in your comment doesn't work with 20 :))

File: config/waagent.logrotate (hunk 5 above)
Comment (vrdmr) [reply config/waagent.logrotate]: Some tests results - 

Uncompressed | Compressed | Compression Ratio
-- | -- | --
3102345801 | 154144402 | 4.97%
393461760 | 19558242 | 4.97%
33047881 | 1770543 | 5.36%
588946964 | 27436410 | 4.66%

Comment (vrdmr) [reply config/waagent.logrotate]: Tested with more chatty agents as well. With the size limit and daily logrotate in place, the files would be rotated every day, once they cross the 20 MB threshold.

```
size (bytes)
   1538167      test-1day.log.gz         # Compressed 1-day log.
  31809193      waagent.log.1-1day.log   # Compressed 1-day log.  (eg: 06/01 06:25:05 - 06/02 06:25:19)
 952814059      waagent.log.1.log        # Uncompressed 1-month log.
```

Comment (larohra) [reply config/waagent.logrotate]: Can we increase the cadence of logrotate to 2-3times a day? If the file size < 20MB it wont do anything and if its >20Mb it'll just rotate it. 
Does the logrotate process use a lot of resources? Any idea?

Comment (vrdmr) [reply config/waagent.logrotate]: The `logrotate` is a common utility and in most cases, `logrotate` is invoked from a script in the `/etc/cron.daily/` directory, which we don't have control over.

And if in the case where it is <20 MB for a long time, it'll remain unrotated - which isnt a problem. We want to catch the cases where we go to 40/50 MB per day usage and compress them immediately. In the previous case, we could go to 950MB for a month, without any rotation.

Comment (larohra) [reply config/waagent.logrotate]: Oh damn I didnt know about the 950mb part, that's insane! 
Thanks for adding it!

**Pull request merged**